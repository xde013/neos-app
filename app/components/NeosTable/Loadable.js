/**
 *
 * Asynchronously loads the component for NeosTable
 *
 */

import loadable from 'utils/loadable';

export default loadable(() => import('./index'));
