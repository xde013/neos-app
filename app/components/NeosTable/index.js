/* eslint-disable react/jsx-props-no-spreading */
/**
 *
 * NeosTable
 *
 */

import React, { forwardRef, useMemo } from 'react';
import PropTypes from 'prop-types';
import MaterialTable from 'material-table';
import Search from '@material-ui/icons/Search';
import Clear from '@material-ui/icons/Clear';
import SaveAlt from '@material-ui/icons/SaveAlt';
import FirstPage from '@material-ui/icons/FirstPage';
import LastPage from '@material-ui/icons/LastPage';
import ChevronLeft from '@material-ui/icons/ChevronLeft';
import ChevronRight from '@material-ui/icons/ChevronRight';
import ArrowDownward from '@material-ui/icons/ArrowDownward';

/**
 * Load only Icons the table needs
 */
const tableIcons = {
  Search: forwardRef((props, ref) => <Search {...props} ref={ref} />),
  Export: forwardRef((props, ref) => <SaveAlt {...props} ref={ref} />),
  ResetSearch: forwardRef((props, ref) => <Clear {...props} ref={ref} />),
  FirstPage: forwardRef((props, ref) => <FirstPage {...props} ref={ref} />),
  LastPage: forwardRef((props, ref) => <LastPage {...props} ref={ref} />),
  NextPage: forwardRef((props, ref) => <ChevronRight {...props} ref={ref} />),
  PreviousPage: forwardRef((props, ref) => (
    <ChevronLeft {...props} ref={ref} />
  )),
  SortArrow: forwardRef((props, ref) => <ArrowDownward {...props} ref={ref} />),
};

export function NeosTable({ neos }) {
  const downloadLinkRef = React.createRef();
  const neosData = useMemo(() => {
    if (neos.length > 0) {
      return neos.map(neo => ({
        name: neo[0],
        estimated_diameter_min: neo[1],
        estimated_diameter_max: neo[2],
      }));
    }
    return [];
  }, [neos]);

  return (
    <div data-testid="table">
      <MaterialTable
        icons={tableIcons}
        title="NEOs Diameters"
        columns={[
          { title: 'Name', field: 'name' },
          {
            title: 'Min Estimated Diameter (km)',
            field: 'estimated_diameter_min',
            type: 'numeric',
            cellStyle: {
              textAlign: 'center',
            },
          },
          {
            title: 'Max Estimated Diameter (km)',
            field: 'estimated_diameter_max',
            type: 'numeric',
            cellStyle: {
              textAlign: 'center',
            },
          },
        ]}
        data={neosData}
        options={{
          exportButton: true,
          exportCsv: columns => {
            const header = `${columns.map(e => e.title).join(',')}\n`;
            const csvString = `${header}${neos
              .map(e => e.join(','))
              .join('\n')}`;
            const blob = new Blob([csvString], {
              type: 'text/csv;charset=utf8;',
            });

            downloadLinkRef.current.setAttribute(
              'href',
              window.URL.createObjectURL(blob),
            );
            downloadLinkRef.current.click();
          },

          headerStyle: {
            backgroundColor: '#0b3d91',
            color: '#FFF',
          },
        }}
      />
      {/* Nicely named download */}
      <a hidden href="#!" download="my_neos_data.csv" ref={downloadLinkRef}>
        Export
      </a>
    </div>
  );
}

NeosTable.propTypes = {
  neos: PropTypes.array.isRequired,
};

export default NeosTable;
