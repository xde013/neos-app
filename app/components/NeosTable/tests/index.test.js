/**
 *
 * Tests for NeosTable
 *
 */

import React from 'react';
import { render } from '@testing-library/react';
import NeosTable from '../index';

describe('<NeosTable />', () => {
  it('Expect to not log errors in console', () => {
    const spy = jest.spyOn(global.console, 'error');
    render(<NeosTable neos={[]} />);
    expect(spy).not.toHaveBeenCalled();
  });

  it('Should render a Table', async () => {
    const { queryByTestId } = render(<NeosTable neos={[]} />);

    expect(queryByTestId(/table/i)).toBeTruthy();
  });

  it('Should render and match the snapshot', () => {
    const {
      container: { firstChild },
    } = render(<NeosTable neos={[]} />);
    expect(firstChild).toMatchSnapshot();
  });
});
