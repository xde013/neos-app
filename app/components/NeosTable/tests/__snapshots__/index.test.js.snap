// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<NeosTable /> Should render and match the snapshot 1`] = `
<div
  data-testid="table"
>
  <div
    class="MuiPaper-root MuiPaper-elevation2 MuiPaper-rounded"
    style="position: relative;"
  >
    <div
      class="MuiToolbar-root MuiToolbar-regular MTableToolbar-root-461 MuiToolbar-gutters"
    >
      <div
        class="MTableToolbar-title-465"
      >
        <h6
          class="MuiTypography-root MuiTypography-h6"
        >
          NEOs Diameters
        </h6>
      </div>
      <div
        class="MTableToolbar-spacer-463"
      />
      <div
        class="MuiFormControl-root MuiTextField-root MTableToolbar-searchField-466"
      >
        <div
          class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-formControl MuiInput-formControl MuiInputBase-adornedStart MuiInputBase-adornedEnd"
        >
          <div
            class="MuiInputAdornment-root MuiInputAdornment-positionStart"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeSmall"
              focusable="false"
              role="presentation"
              title="Search"
              viewBox="0 0 24 24"
            >
              <path
                d="M15.5 14h-.79l-.28-.27C15.41 12.59 16 11.11 16 9.5 16 5.91 13.09 3 9.5 3S3 5.91 3 9.5 5.91 16 9.5 16c1.61 0 3.09-.59 4.23-1.57l.27.28v.79l5 4.99L20.49 19l-4.99-5zm-6 0C7.01 14 5 11.99 5 9.5S7.01 5 9.5 5 14 7.01 14 9.5 11.99 14 9.5 14z"
              />
            </svg>
          </div>
          <input
            aria-invalid="false"
            class="MuiInputBase-input MuiInput-input MuiInputBase-inputAdornedStart MuiInputBase-inputAdornedEnd"
            placeholder="Search"
            type="text"
            value=""
          />
          <div
            class="MuiInputAdornment-root MuiInputAdornment-positionEnd"
          >
            <button
              class="MuiButtonBase-root MuiIconButton-root Mui-disabled Mui-disabled"
              disabled=""
              tabindex="-1"
              type="button"
            >
              <span
                class="MuiIconButton-label"
              >
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root MuiSvgIcon-fontSizeSmall"
                  focusable="false"
                  role="presentation"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M19 6.41L17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
                  />
                </svg>
              </span>
            </button>
          </div>
        </div>
      </div>
      <div
        class="MTableToolbar-actions-464"
      >
        <div>
          <div>
            <span>
              <button
                aria-label="Export"
                class="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorInherit"
                tabindex="0"
                title="Export"
                type="button"
              >
                <span
                  class="MuiIconButton-label"
                >
                  <svg
                    aria-hidden="true"
                    class="MuiSvgIcon-root"
                    focusable="false"
                    role="presentation"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M19 12v7H5v-7H3v7c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2v-7h-2zm-6 .67l2.59-2.58L17 11.5l-5 5-5-5 1.41-1.41L11 12.67V3h2z"
                    />
                  </svg>
                </span>
                <span
                  class="MuiTouchRipple-root"
                />
              </button>
            </span>
            <span />
          </div>
        </div>
      </div>
    </div>
    <div
      style="overflow-x: auto;"
    >
      <div>
        <div
          style="overflow-y: auto;"
        >
          <table
            class="MuiTable-root"
          >
            <thead
              class="MuiTableHead-root"
            >
              <tr
                class="MuiTableRow-root MuiTableRow-head"
              >
                <th
                  class="MuiTableCell-root MuiTableCell-head MTableHeader-header-586 MuiTableCell-alignLeft"
                  scope="col"
                  style="background-color: rgb(11, 61, 145); color: rgb(255, 255, 255);"
                >
                  <span
                    aria-disabled="false"
                    class="MuiButtonBase-root MuiTableSortLabel-root"
                    role="button"
                    tabindex="0"
                  >
                    <div
                      aria-roledescription="Draggable item. Press space bar to lift"
                      data-react-beautiful-dnd-drag-handle="2"
                      data-react-beautiful-dnd-draggable="2"
                      draggable="false"
                      tabindex="0"
                    >
                      Name
                    </div>
                    <svg
                      aria-hidden="true"
                      class="MuiSvgIcon-root MuiTableSortLabel-icon MuiTableSortLabel-iconDirectionAsc"
                      focusable="false"
                      role="presentation"
                      viewBox="0 0 24 24"
                    >
                      <path
                        d="M20 12l-1.41-1.41L13 16.17V4h-2v12.17l-5.58-5.59L4 12l8 8 8-8z"
                      />
                    </svg>
                  </span>
                </th>
                <th
                  class="MuiTableCell-root MuiTableCell-head MTableHeader-header-586 MuiTableCell-alignRight"
                  scope="col"
                  style="background-color: rgb(11, 61, 145); color: rgb(255, 255, 255);"
                >
                  <span
                    aria-disabled="false"
                    class="MuiButtonBase-root MuiTableSortLabel-root"
                    role="button"
                    tabindex="0"
                  >
                    <div
                      aria-roledescription="Draggable item. Press space bar to lift"
                      data-react-beautiful-dnd-drag-handle="2"
                      data-react-beautiful-dnd-draggable="2"
                      draggable="false"
                      tabindex="0"
                    >
                      Min Estimated Diameter (km)
                    </div>
                    <svg
                      aria-hidden="true"
                      class="MuiSvgIcon-root MuiTableSortLabel-icon MuiTableSortLabel-iconDirectionAsc"
                      focusable="false"
                      role="presentation"
                      viewBox="0 0 24 24"
                    >
                      <path
                        d="M20 12l-1.41-1.41L13 16.17V4h-2v12.17l-5.58-5.59L4 12l8 8 8-8z"
                      />
                    </svg>
                  </span>
                </th>
                <th
                  class="MuiTableCell-root MuiTableCell-head MTableHeader-header-586 MuiTableCell-alignRight"
                  scope="col"
                  style="background-color: rgb(11, 61, 145); color: rgb(255, 255, 255);"
                >
                  <span
                    aria-disabled="false"
                    class="MuiButtonBase-root MuiTableSortLabel-root"
                    role="button"
                    tabindex="0"
                  >
                    <div
                      aria-roledescription="Draggable item. Press space bar to lift"
                      data-react-beautiful-dnd-drag-handle="2"
                      data-react-beautiful-dnd-draggable="2"
                      draggable="false"
                      tabindex="0"
                    >
                      Max Estimated Diameter (km)
                    </div>
                    <svg
                      aria-hidden="true"
                      class="MuiSvgIcon-root MuiTableSortLabel-icon MuiTableSortLabel-iconDirectionAsc"
                      focusable="false"
                      role="presentation"
                      viewBox="0 0 24 24"
                    >
                      <path
                        d="M20 12l-1.41-1.41L13 16.17V4h-2v12.17l-5.58-5.59L4 12l8 8 8-8z"
                      />
                    </svg>
                  </span>
                </th>
              </tr>
            </thead>
            <tbody
              class="MuiTableBody-root"
            >
              <tr
                class="MuiTableRow-root"
                style="height: 245px;"
              >
                <td
                  class="MuiTableCell-root MuiTableCell-body"
                  colspan="3"
                  style="padding-top: 0px; padding-bottom: 0px; text-align: center;"
                >
                  No records to display
                </td>
              </tr>
            </tbody>
          </table>
        </div>
      </div>
    </div>
    <table
      class="MuiTable-root"
    >
      <tfoot
        class="MuiTableFooter-root"
        style="display: grid;"
      >
        <tr
          class="MuiTableRow-root MuiTableRow-footer"
        >
          <td
            class="MuiTableCell-root MuiTableCell-footer MuiTablePagination-root Component-paginationRoot-429"
            colspan="3"
            icons="[object Object]"
            style="float: right; overflow-x: auto;"
          >
            <div
              class="MuiToolbar-root MuiToolbar-regular MuiTablePagination-toolbar Component-paginationToolbar-430 MuiToolbar-gutters"
            >
              <div
                class="MuiTablePagination-spacer"
              />
              <p
                class="MuiTypography-root MuiTablePagination-caption Component-paginationCaption-431 MuiTypography-body2 MuiTypography-colorInherit"
              >
                Rows per page:
              </p>
              <div
                class="MuiInputBase-root MuiTablePagination-input MuiTablePagination-selectRoot Component-paginationSelectRoot-432"
              >
                <div
                  aria-haspopup="listbox"
                  aria-labelledby=" "
                  class="MuiSelect-root MuiSelect-select MuiTablePagination-select MuiSelect-selectMenu MuiInputBase-input"
                  role="button"
                  tabindex="0"
                >
                  <div
                    style="padding: 0px 5px;"
                  >
                    5 rows 
                  </div>
                </div>
                <input
                  type="hidden"
                  value="5"
                />
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root MuiSelect-icon MuiTablePagination-selectIcon"
                  focusable="false"
                  role="presentation"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M7 10l5 5 5-5z"
                  />
                </svg>
              </div>
              <p
                class="MuiTypography-root MuiTablePagination-caption Component-paginationCaption-431 MuiTypography-body2 MuiTypography-colorInherit"
              >
                0-0 of 0
              </p>
              <div
                class="MTablePaginationInner-root-632"
              >
                <span
                  class=""
                  title="First Page"
                >
                  <button
                    class="MuiButtonBase-root MuiIconButton-root Mui-disabled Mui-disabled"
                    disabled=""
                    tabindex="-1"
                    type="button"
                  >
                    <span
                      class="MuiIconButton-label"
                    >
                      <svg
                        aria-hidden="true"
                        class="MuiSvgIcon-root"
                        focusable="false"
                        role="presentation"
                        viewBox="0 0 24 24"
                      >
                        <path
                          d="M18.41 16.59L13.82 12l4.59-4.59L17 6l-6 6 6 6zM6 6h2v12H6z"
                        />
                        <path
                          d="M24 24H0V0h24v24z"
                          fill="none"
                        />
                      </svg>
                    </span>
                  </button>
                </span>
                <span
                  class=""
                  title="Previous Page"
                >
                  <button
                    class="MuiButtonBase-root MuiIconButton-root Mui-disabled Mui-disabled"
                    disabled=""
                    tabindex="-1"
                    type="button"
                  >
                    <span
                      class="MuiIconButton-label"
                    >
                      <svg
                        aria-hidden="true"
                        class="MuiSvgIcon-root"
                        focusable="false"
                        role="presentation"
                        viewBox="0 0 24 24"
                      >
                        <path
                          d="M15.41 7.41L14 6l-6 6 6 6 1.41-1.41L10.83 12z"
                        />
                      </svg>
                    </span>
                  </button>
                </span>
                <span
                  class="MuiTypography-root MuiTypography-caption"
                  style="flex: 1; text-align: center; align-self: center; flex-basis: inherit;"
                >
                  1-0 of 0
                </span>
                <span
                  class=""
                  title="Next Page"
                >
                  <button
                    class="MuiButtonBase-root MuiIconButton-root Mui-disabled Mui-disabled"
                    disabled=""
                    tabindex="-1"
                    type="button"
                  >
                    <span
                      class="MuiIconButton-label"
                    >
                      <svg
                        aria-hidden="true"
                        class="MuiSvgIcon-root"
                        focusable="false"
                        role="presentation"
                        viewBox="0 0 24 24"
                      >
                        <path
                          d="M10 6L8.59 7.41 13.17 12l-4.58 4.59L10 18l6-6z"
                        />
                      </svg>
                    </span>
                  </button>
                </span>
                <span
                  class=""
                  title="Last Page"
                >
                  <button
                    class="MuiButtonBase-root MuiIconButton-root Mui-disabled Mui-disabled"
                    disabled=""
                    tabindex="-1"
                    type="button"
                  >
                    <span
                      class="MuiIconButton-label"
                    >
                      <svg
                        aria-hidden="true"
                        class="MuiSvgIcon-root"
                        focusable="false"
                        role="presentation"
                        viewBox="0 0 24 24"
                      >
                        <path
                          d="M5.59 7.41L10.18 12l-4.59 4.59L7 18l6-6-6-6zM16 6h2v12h-2z"
                        />
                      </svg>
                    </span>
                  </button>
                </span>
              </div>
            </div>
          </td>
        </tr>
      </tfoot>
    </table>
  </div>
  <a
    download="my_neos_data.csv"
    hidden=""
    href="#!"
  >
    Export
  </a>
</div>
`;
